src/python/pants/backend/openapi/util_rules/openapi_document.py

```diff
@@ -0,0 +1,116 @@
+# Copyright 2024 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+from pathlib import PurePath
+
+from pants.backend.javascript.subsystems import nodejs_tool
+from pants.backend.javascript.subsystems.nodejs_tool import NodeJSToolRequest
+from pants.backend.openapi.subsystems.redocly import Redocly
+from pants.backend.openapi.target_types import (
+    OpenApiDocumentField,
+    OpenApiDocumentGeneratorTarget,
+    OpenApiDocumentTarget,
+    OpenApiSourceField,
+)
+from pants.core.target_types import ResourceSourceField
+from pants.core.util_rules.source_files import SourceFilesRequest
+from pants.core.util_rules.stripped_source_files import StrippedSourceFiles
+from pants.engine.internals.native_engine import AddPrefix, Digest, Snapshot
+from pants.engine.internals.selectors import Get, MultiGet
+from pants.engine.process import ProcessResult
+from pants.engine.rules import collect_rules, rule
+from pants.engine.target import (
+    GeneratedSources,
+    GenerateSourcesRequest,
+    StringField,
+    TransitiveTargets,
+    TransitiveTargetsRequest,
+)
+from pants.engine.unions import UnionRule
+from pants.source.source_root import SourceRoot, SourceRootRequest
+from pants.util.logging import LogLevel
+from pants.util.strutil import help_text, pluralize
+
+
+class BundleOpenApiDocumentRequest(GenerateSourcesRequest):
+    input = OpenApiDocumentField
+    output = ResourceSourceField
+
+
+class BundleSourceRootField(StringField):
+    alias = "bundle_source_root"
+    help = help_text(
+        """
+        The source root to bundle OpenAPI documents under.
+
+        If unspecified, the source root the `openapi_documents` is under will be used.
+        """
+    )
+
+
+@rule
+async def bundle_openapi_document(
+    request: BundleOpenApiDocumentRequest, redocly: Redocly
+) -> GeneratedSources:
+    target = request.protocol_target
+    bundle_source_root = request.protocol_target.get(BundleSourceRootField).value
+    transitive_targets = await Get(TransitiveTargets, TransitiveTargetsRequest([target.address]))
+    source_root_request = Get(
+        SourceRoot,
+        SourceRootRequest,
+        SourceRootRequest(PurePath(bundle_source_root))
+        if bundle_source_root
+        else SourceRootRequest.for_target(request.protocol_target),
+    )
+
+    target_stripped_sources_request = Get(
+        StrippedSourceFiles, SourceFilesRequest([target[OpenApiDocumentField]])
+    )
+    all_stripped_sources_request = Get(
+        StrippedSourceFiles,
+        SourceFilesRequest(
```

Comment:
Could you describe why the source root stripping is necessary? Without knowing anything about the tools, I would've imagined `redocly bundle src/openapi/foo/openapi.json -o src/openapi/foo/openapi.json` would do approximately the right thing?

Reply:
Yes, it would. I'm stripping the source root because it can optionally be changed with `bundle_source_root` similar to protobuf's `python_source_root`. Or at least I assumed it's why the Python protobuf codegen stripped it, maybe it does it for another reason and the source root can be changed post-process anyway :)

Reply:
Oh, I see okay, so the bundled document can be placed somewhere else. Makes sense!